File: lab5/Haskell/math.hs
-- Haskell ❤️❤️❤️

binomial :: Int -> Int -> Int
binomial n 0 = 1
binomial n k = if n == k then 1
               else binomial (pred n) k + binomial (pred n) (pred k)

build_triangle_row :: Int -> [Int] -> [Int]
build_triangle_row 0 prev_row = prev_row
build_triangle_row n prev_row = build_triangle_row (pred n) next_row
                                where
                                    next_row = 1 : zipWith (+) (init prev_row) (tail prev_row) ++ [1]
                                
binomial2 :: Int -> Int -> Int
binomial2 n 0 = 1
binomial2 n k = if n == k then 1
                else (build_triangle_row n [1])!!k

merge :: [Int] -> [Int] -> [Int]
merge [] right_part = right_part
merge left_part []  = left_part
merge left_part right_part = if head left_part < head right_part then 
                                head left_part : (merge (tail left_part) right_part)
                             else 
                                head right_part : (merge left_part (tail right_part))

merge_sort :: [Int] -> [Int]
merge_sort [] = []
merge_sort [x] = [x]
merge_sort list = merge (merge_sort left_part) (merge_sort right_part)
                  where
                        split_result = splitAt (div (length list) 2) list
                        left_part = fst split_result
                        right_part = snd split_result
         
de :: Int -> Int -> (Int, Int, Int)
de 0 b = (0, 1, b)
de a 0 = (1, 0, a)
de a b = (x, y, gcd)
         where
            (x1, y1, gcd) = de b (mod a b)
            x = y1
            y =  x1 - (div a b) * y1

is_prime :: Int -> Bool
is_prime 2 = True
is_prime 3 = True
is_prime n
            | n < 2 = False 
            | otherwise = not (any (\x -> mod n x == 0) [2..isqrt n])
                where
                isqrt :: Int -> Int
                isqrt = floor . sqrt . fromIntegral

get_next_prime :: Int -> Int
get_next_prime 2 = 2
get_next_prime 3 = 3
get_next_prime n = if is_prime n then n
                   else
                       get_next_prime (succ n)


get_prime_factors :: Int -> Int -> [Int]
get_prime_factors n start = if next_prime > n then
                        []
                      else 
                        if mod n next_prime == 0 then
                            next_prime : get_prime_factors next_n start
                        else
                            get_prime_factors n (next_prime + 1)
                      where
                        next_prime = get_next_prime start
                        next_n = div n next_prime

prime_factors :: Int -> [Int]
prime_factors n = get_prime_factors n 2

totient :: Int -> Int
totient n = length([x | x <- [1..n], gcd x n == 1])
 
totient2 :: Int -> Int
totient2 n = length([ x | x <- [1..n], not (any (\y -> mod x y == 0) (prime_factors n) ) ])

primes :: Int -> [Int]
primes n | n < 2 = []
         | otherwise = [x | x <- [2..n], is_prime x == True]

File: lab5/Prolog/math.pl
merge([], RIGHT, RIGHT).
merge(LEFT, [], LEFT).
merge([HL|TL], [HR|TR], [HL|T]) :-
        HL =< HR,
        merge(TL, [HR|TR], T).
merge([HL|TL], [HR|TR], [HR|T]) :-
        HL > HR,
        merge([HL|TL],TR, T).

split([], [], []).
split([X], [X], []).
split([H1, H2|T], [H1|T1], [H2|T2]) :-
    split(T, T1, T2).

mergesort([], []).
mergesort([X], [X]).
mergesort(List, Sorted) :-
    split(List, L1, L2),
    mergesort(L1, Sorted1),
    mergesort(L2, Sorted2),
    merge(Sorted1, Sorted2, Sorted).


gcd(0, B, B) :- B \= 0.
gcd(A, 0, A) :- A \= 0.
gcd(A, B, G) :-
    A \= 0,
    B \= 0,
    R is A mod B,
    gcd(B, R, G).

de(0, B, 0, 1, B) :- B \= 0.
de(A, 0, 1, 0, A) :- A \= 0.
de(A, B, X, Y, G) :-
    B \= 0,
    Q is A // B,
    R is A mod B,
    de(B, R, X1, Y1, G),
    X is Y1,
    Y is X1 - Q * Y1.


is_prime(2).
is_prime(3).
is_prime(N) :-
    N > 3,
    N mod 2 =\= 0,
    \+ has_factor(N, 3).

has_factor(N, F) :-
    N mod F =:= 0.
has_factor(N, F) :-
    F * F < N,
    F2 is F + 2,
    has_factor(N, F2).

next_prime(P, NP) :-
    NP is P + 1,
    is_prime(NP).

smallest_prime_factor(N, F) :-
    between(2, N, F),
    N mod F =:= 0,
    is_prime(F), !.

prime_factors(1, []).
prime_factors(N, [F|Fs]) :-
    N > 1,
    smallest_prime_factor(N, F),
    N1 is N // F,
    prime_factors(N1, Fs).


totient(N, T) :-
    findall(X, (between(1, N, X), gcd(N, X, 1)), RelPrimes),
    length(RelPrimes, T).

primes(N, P) :-
    findall(X, (between(2, N, X), 
    is_prime(X)), P).